File: 1572.矩阵对角线元素的和.cpp
/*
 * @lc app=leetcode.cn id=1572 lang=cpp
 *
 * [1572] 矩阵对角线元素的和
 */

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    int diagonalSum(vector<vector<int>>& mat) {
        int n = mat.size(), ret = 0;
        for(int i=0; i<n; ++i){
            ret += mat[i][i];
            ret += mat[i][n-i-1];
        }
        if(n & 1){
            ret -= mat[n/2][n/2];
        }
        return ret;
    }
};
// @lc code=end

File: 1572.矩阵对角线元素的和.py
#
# @lc app=leetcode.cn id=1572 lang=python3
#
# [1572] 矩阵对角线元素的和
#

# @lc code=start
from typing import List
import numpy as np

class Solution:
    def diagonalSum(self, mat: List[List[int]]) -> int:
        matrix = np.array(mat)
        n = matrix.shape[0]
        return int(matrix[range(n), range(n)].sum() + 
            matrix[range(n), range(n-1, -1, -1)].sum() - 
            (matrix[(n-1)//2:n//2+1, (n-1)//2:n//2+1].sum() if n & 1 else 0)
        )
# @lc code=end

print(Solution().diagonalSum(mat = [
            [1,2,3],
            [4,5,6],
            [7,8,9]]))
print(Solution().diagonalSum(mat = [[1,1,1,1],
            [1,1,1,1],
            [1,1,1,1],
            [1,1,1,1]]))
print(Solution().diagonalSum(mat = [[5]]))